Assert failure: Thread::IsObjRefValid(&objref) testing checked against corefx

Windows_NT x64 Checked corefx_baseline failure
```
  Assert failure(PID 6704 [0x00001a30], Thread: 4120 [0x1018]): Thread::IsObjRefValid(&objref)
19:18:16   
19:18:16   CORECLR! OBJECTREF::OBJECTREF + 0x11D (0x00007ffd`63f1505d)
19:18:16   CORECLR! ExceptionNative::GetStackTracesDeepCopy + 0x490 (0x00007ffd`64275a80)
19:18:16   SYSTEM.PRIVATE.CORELIB! <no symbol> + 0x0 (0x00007ffd`6389c6f7)
19:18:16   SYSTEM.PRIVATE.CORELIB! <no symbol> + 0x0 (0x00007ffd`6389c7ac)
19:18:16   <no module>! <no symbol> + 0x0 (0x00007ffd`0596fd2d)
19:18:16   <no module>! <no symbol> + 0x0 (0xffffffff`fffffffe)
19:18:16   <no module>! <no symbol> + 0x0 (0x000000fc`4589b450)
19:18:16   <no module>! <no symbol> + 0x0 (0x00000000`00000001)
19:18:16   CORECLR! GetClrDebugState + 0x23 (0x00007ffd`63e38c8f)
19:18:16   CORECLR! ENABLESTRESSHEAP + 0xE (0x00007ffd`63f188de)
19:18:16       File: d:\j\workspace\x64_checked_w---d7295605\src\vm\object.cpp Line: 2662
19:18:16       Image: D:\j\workspace\x64_checked_w---d7295605\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-x64\dotnet.exe
19:18:16   
19:18:16   Finished running tests.  End time=19:18:16.96, Exit code = 123456789
19:18:16 D:\j\workspace\x64_checked_w---d7295605\_\fx\Tools\tests.targets(484,5): 
warning MSB3073: The command "D:\j\workspace\x64_checked_w---d7295605\_\fx\bin/tests/System.Linq.Expressions.Tests/netcoreapp-Windows_NT-Release-x64//RunTests.cmd 
D:\j\workspace\x64_checked_w---d7295605\_\fx\bin/testhost/netcoreapp-Windows_NT-Release-x64/" 
exited with code 123456789. 
[D:\j\workspace\x64_checked_w---d7295605\_\fx\src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj]
```

/cc @jkotas

--- a/src/vm/object.cpp
+++ b/src/vm/object.cpp
@@ -2997,6 +2997,7 @@ void StackTraceArray::Append(StackTraceElement const * begin, StackTraceElement 
         THROWS;
         GC_TRIGGERS;
         MODE_COOPERATIVE;
+        PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)this));
     }
     CONTRACTL_END;
 
@@ -3021,6 +3022,7 @@ void StackTraceArray::AppendSkipLast(StackTraceElement const * begin, StackTrace
         THROWS;
         GC_TRIGGERS;
         MODE_COOPERATIVE;
+        PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)this));
     }
     CONTRACTL_END;
 
@@ -3048,8 +3050,9 @@ void StackTraceArray::AppendSkipLast(StackTraceElement const * begin, StackTrace
     else
     {
         // slow path: create a copy and append
-        StackTraceArray copy(*this);
+        StackTraceArray copy;
         GCPROTECT_BEGIN(copy);
+            copy.CopyFrom(*this);
             copy.SetSize(copy.Size() - 1);
             copy.Append(begin, end);
             this->Swap(copy);
@@ -3091,6 +3094,7 @@ void StackTraceArray::Grow(size_t grow_size)
         GC_TRIGGERS;
         MODE_COOPERATIVE;
         INJECT_FAULT(ThrowOutOfMemory(););
+        PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)this));
     }
     CONTRACTL_END;
 
@@ -3132,37 +3136,17 @@ void StackTraceArray::EnsureThreadAffinity()
     {
         // object is being changed by a thread different from the one which created it
         // make a copy of the array to prevent a race condition when two different threads try to change it
-        StackTraceArray copy(*this);
-        this->Swap(copy);
+        StackTraceArray copy;
+        GCPROTECT_BEGIN(copy);
+            copy.CopyFrom(*this);
+            this->Swap(copy);
+        GCPROTECT_END();
     }
 }
 
 #ifdef _MSC_VER
 #pragma warning(disable: 4267) 
 #endif
-
-StackTraceArray::StackTraceArray(StackTraceArray const & rhs)
-{
-    CONTRACTL
-    {
-        THROWS;
-        GC_TRIGGERS;
-        MODE_COOPERATIVE;
-        INJECT_FAULT(ThrowOutOfMemory(););
-    }
-    CONTRACTL_END;
-
-    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(rhs.Capacity()));
-
-    GCPROTECT_BEGIN(m_array);
-        Volatile<size_t> size = rhs.Size();
-        memcpyNoGCRefs(GetRaw(), rhs.GetRaw(), size * sizeof(StackTraceElement) + sizeof(ArrayHeader));
-
-        SetSize(size);  // set size to the exact value which was used when we copied the data
-                        // another thread might have changed it at the time of copying
-        SetObjectThread();  // affinitize the newly created array with the current thread
-    GCPROTECT_END();
-}
 
 // Deep copies the stack trace array
 void StackTraceArray::CopyFrom(StackTraceArray const & src)
@@ -3173,19 +3157,19 @@ void StackTraceArray::CopyFrom(StackTraceArray const & src)
         GC_TRIGGERS;
         MODE_COOPERATIVE;
         INJECT_FAULT(ThrowOutOfMemory(););
+        PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)this));
+        PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)&src));
     }
     CONTRACTL_END;
 
     m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Capacity()));
 
-    GCPROTECT_BEGIN(m_array);
     Volatile<size_t> size = src.Size();
     memcpyNoGCRefs(GetRaw(), src.GetRaw(), size * sizeof(StackTraceElement) + sizeof(ArrayHeader));
 
     SetSize(size);  // set size to the exact value which was used when we copied the data
                     // another thread might have changed it at the time of copying
     SetObjectThread();  // affinitize the newly created array with the current thread
-    GCPROTECT_END();
 }
 
 #ifdef _MSC_VER

--- a/src/vm/object.h
+++ b/src/vm/object.h
@@ -3110,15 +3110,9 @@ public:
     void CopyFrom(StackTraceArray const & src);
     
 private:
-    StackTraceArray(StackTraceArray const & rhs);
+    StackTraceArray(StackTraceArray const & rhs) = delete;
 
-    StackTraceArray & operator=(StackTraceArray const & rhs)
-    {
-        WRAPPER_NO_CONTRACT;
-        StackTraceArray copy(rhs);
-        this->Swap(copy);
-        return *this;
-    }
+    StackTraceArray & operator=(StackTraceArray const & rhs) = delete;
 
     void Grow(size_t size);
     void EnsureThreadAffinity();

